### PR TITLE
Fixes and Logging

### PR DESCRIPTION
**Pull request by ydylla**
Hi,
my experiments mentioned in #68 revealed some bugs. I also added some more logging which makes debugging configs a lot easier.  

The first bug (at least I think it is one :smile:) is that when a matcher returns an error the processing is not stopped. Instead the next matcher is tried and if that matches its handlers are executed even if the connection is broken.

For example with this config
<details>
<summary>matcher-bug.json</summary>

```json
{
  "admin": {
    "disabled": true
  },
  "logging": {
    "logs": {
      "default": {"level":"DEBUG", "encoder": {"format":"console"}}
    }
  },
  "apps": {
    "layer4": {
      "servers": {
        "matcher": {
          "listen": ["0.0.0.0:8888"],
          "routes": [
            {
              "match": [
                {
                  "tls": {}
                }
              ],
              "handle": [
                {
                  "handler": "echo"
                }
              ]
            },
            {
              "match": [
                {
                  "ip": {"ranges": ["0.0.0.0/0"]}
                }
              ],
              "handle": [
                {
                  "handler": "proxy",
                  "upstreams": [{"dial": ["127.0.0.1:9999"]}]
                }
              ]
            }
          ]
        }
      }
    }
  }
}
```

</details>

and this curl `curl -x socks5://127.0.0.1:8888 https://example.org` (Ctrl-C or wait 5 minutes for the timeout) you will get this confusing log:

```
1.6627550161737518e+09  info    serving initial configuration
1.662755029922435e+09   error   layer4  matching connection     {"error": "unexpected EOF"}
1.6627550319487917e+09  debug   layer4.handlers.proxy   dial upstream   {"remote": "10.0.0.2:43900", "upstream": "127.0.0.1:9999", "error": "dial tcp 127.0.0.1:9999: connectex: Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte."}
1.662755031949321e+09   error   layer4  handling connection     {"error": "dial tcp 127.0.0.1:9999: connectex: Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte."}
1.662755031949321e+09   debug   layer4  connection stats        {"remote": "10.0.0.2:43900", "read": 4, "written": 0, "duration": 6.8051304}
```

with the fixes and additional logging it looks like this:
```
1.6627555421950946e+09  info    serving initial configuration
1.6627555549489841e+09  debug   layer4  matching        {"remote": "10.0.0.2:35720", "error": "unexpected EOF", "matcher": "layer4.matchers.tls", "matched": false}
1.6627555549492755e+09  error   layer4  matching connection     {"remote": "10.0.0.2:35720", "error": "unexpected EOF"}
1.6627555549492755e+09  debug   layer4  connection stats        {"remote": "10.0.0.2:35720", "read": 4, "written": 0, "duration": 3.2630775}
```
Only one matcher and no handlers are executed.

The second bug is causing `Read()` to return 0 bytes and no error on its first call after `record()` (which I also mentioned [here](https://github.com/mholt/caddy-l4/pull/68#issuecomment-1239907686)). It's probably not forbidden for an `io.Reader` to do that but its confusing. I wrote a connection test that covers this situation `TestConnection_RecordAndRewind`.

The additional remote ip address logging is mainly intended as kind of connection id, with that its possible to correlate multiple log lines on busy servers.

Besides that I also have a commit which adds a per route configurable matching timeout. But it needs more testing and is probably better discussed in its own PR.

I am also still experimenting on how to prevent matchers from blocking when there is not enough data for them. But I would like to see these fixes merged first, so I can work on top of them. A short read detection in the layer4 connection looks promising. One problem is the http matcher. Since it uses a buffered reader, it will always cause a short read from the view point of the connection.

### CODE REVIEW

**Comment by ydylla** [comment]
During testing with a real config I noticed that my 0 bytes fix was flawed, I force pushed a better one.

**Review by mholt — COMMENTED** [review]
Again, just did a brief review for now, but it's looking pretty good so far.

I wonder if we can use the MatcherSet without having to pass in a logger, though. Hmm.

File: layer4/matchers.go
Comment (ydylla) [reply layer4/matchers.go:32]: To make sure that `CaddyModule()` can be called in the [Match](https://github.com/mholt/caddy-l4/pull/71/files#diff-df4c4d5d1b7d95094092aea25682cc4da237cc69096bf62f2497a9285c070e86R55) function to extract the matcher name.

File: layer4/matchers.go
Comment (ydylla) [reply layer4/matchers.go:19]: Interesting I had to manually remove the blank line below. Otherwise, nothing changed on `go fmt`.

File: layer4/connection.go
Comment (ydylla) [reply layer4/connection.go:88]: It is just unnecessary, instead we can easily advance to a real Read. It was annoying during debugging and also interfered with `TestConnection_RecordAndRewind`. So yes not really a bug, but it makes things easier.

**Comment by ydylla** [comment]
> I wonder if we can use the MatcherSet without having to pass in a logger
Would you prefer it if `layer4.Connection` had a public logger? I already did this in one of my experiment branches.  
Or did you mean an additional `nil` check before logger usage? Logging which matchers are executed is one of the main objectives of this PR 😄

File: layer4/connection.go
Comment (ydylla) [reply layer4/connection.go:88]: To be clear the 0 Read always happens at the first Read because how we setup the `bufReader` in `record`.

File: layer4/matchers.go
Comment (mholt) [reply layer4/matchers.go:32]: Ah, now I see it.

I wonder, if that could be optional. It doesn't _strictly_ have to be a Caddy module (though almost always will be!) and since it's just for logging I wonder if a type assertion could be used to optionally get that information. That might be better than requiring another interface here.

File: layer4/connection.go
Comment (mholt) [reply layer4/connection.go:88]: Gotcha. That's fine then. I might suggest that the comment be moved to its own line and say something more specific:

```
// first read can be empty due to bufReader
```

or something like that?

Or, should we adjust `bufReader` to not return an empty read instead?

**Comment by mholt** [comment]
> Would you prefer it if layer4.Connection had a public logger?

@ydylla Maybe that's a good idea! Can we try it?

File: layer4/matchers.go
Comment (ydylla) [reply layer4/matchers.go:32]: Yes I can change this. But what should the logged matcher name be if the type assertion fails? Just `unknown` ? Without the `matcher` field the log message barely makes sense but omitting it would be even worse for debugging.

File: layer4/connection.go
Comment (ydylla) [reply layer4/connection.go:88]: I do not see an easy to understand way to init `bufReader` which prevents the 0 Read. The if condition before the return is easier to understand in my opinion, since its affect is more local.

**Comment by ydylla** [comment]
> Maybe that's a good idea! Can we try it?

Sure, I will change it tomorrow.

File: layer4/matchers.go
Comment (mholt) [reply layer4/matchers.go:32]: @ydylla You can use the `.With()` method to conditionally add the field, or even better yet, I think the zap logger omits serializing empty field values. (I know it omits nil, not sure about empty.)

Comment (ydylla) [reply layer4/matchers.go:32]: Yes but I think reading just:
`debug layer4 matching {"remote": "10.0.0.2:35720", "matched": false}`
is not very helpful. I would prefer:
`debug layer4 matching {"remote": "10.0.0.2:35720", "matcher": "unknown", "matched": false}`
which makes it clear the name could not be extracted and one maybe want to look into that as well.

**Comment by ydylla** [comment]
@mholt I have added a logger to the layer4 connection and also removed the `caddy.Module` requirement for `ConnMatcher`.

File: layer4/routes.go
Comment (mholt) [inline layer4/routes.go:131]: Where would it be logged again? I don't see anywhere else we use the value `err`.

Comment (ydylla) [reply layer4/routes.go:131]: It would be logged again in [server.go#L109](https://github.com/mholt/caddy-l4/blob/acf0ea80461d5931e927fc1e9b3d186655a62417/layer4/server.go#L109)

**Comment by mholt** [comment]
(Sorry, still working through my backlog. I'm trying to get Caddy 2.6 ready to release. Your PRs are the next on my list. Thanks for your patience!)

File: layer4/routes.go
Comment (mholt) [reply layer4/routes.go:131]: Interesting, are you seeing that in your logs? How does `err` get returned from here?

Comment (ydylla) [reply layer4/routes.go:131]: The return statement is completely new to fix this:
> The first bug (at least I think it is one 😄) is that when a matcher returns an error the processing is not stopped. Instead the next matcher is tried and if that matches its handlers are executed even if the connection is broken.

You can see example logs of the processing not stopping in the initial post. When using `return err` instead one would see the double logging.

**Comment by ydylla** [comment]
Hi @mholt,
I am still interested to see this merged. What is missing here to move this forward?

File: layer4/routes.go
Comment (mholt) [reply layer4/routes.go:131]: I follow so far, except I don't see where we are using `return err` anywhere. We always return `nextCopy.Handle(cx)` instead.

(Sorry if I'm being a bit slow to understand.)

Comment (ydylla) [reply layer4/routes.go:131]: I think we are talking past each other :)

I included the comment to signify we are returning `nil` on purpose, because otherwise it may look like a bug.  

You are right we mostly return `nextCopy.Handle(cx)` except for exactly this case, since we want to stop the execution of the chain. But we already logged (handled) the error, so it is wrong to return the error (like it's often the case in `if err != nil` blocks). Because the last return value from the chain will eventually appear in [server.go#L106](https://github.com/mholt/caddy-l4/blob/acf0ea80461d5931e927fc1e9b3d186655a62417/layer4/server.go#L106) and thus would get logged in [server.go#L109](https://github.com/mholt/caddy-l4/blob/acf0ea80461d5931e927fc1e9b3d186655a62417/layer4/server.go#L109) which we do not want because we already logged it.

Comment (ydylla) [reply layer4/routes.go:131]: Hi @mholt, did you found the time to think about this?

Comment (mholt) [reply layer4/routes.go:131]: @ydylla I'm so sorry for the delay on this PR. (It's not fair to you, that's for sure.) I was traveling this last week and also taking a little break from sprinting on Caddy stuff after the 2.6 release (to avoid burnout), and now I'm way behind on notifications. :sweat_smile: I'll be catching up soon!

If we can find another maintainer who is willing to review and approve this and other PRs, that might be a good idea so it's not always blocked by me. I feel bad about that! I might look around and see who would be willing to accept the invitation. This plugin just isn't nearly as active as the main Caddy project so it has fewer candidates.

Comment (ydylla) [reply layer4/routes.go:131]: No problem at all, I understand that you are very busy. Don't feel bad, you are doing good work with caddy.  
But you have to endure me pining you from time to time 😄 

I just used the delay to clean up the commits and fix the failing tests in the l4socks module.

**Review by mholt — APPROVED** [review]
@ydylla I suppose you have tested this more than I will be able to at the moment. I'm willing to merge it in!

Fire away when ready.

(Last night I asked in your other PR why a separate Logger was added for each connection. I guess that is being introduced here, which I now I remember. Is having a separate logger for each connection particularly useful?)

Thanks for all you do :)

**Comment by ydylla** [comment]
@mholt Thanks, I will merge it when the logger stuff is settled.  
Yes #72 is based on this branch.

> Is having a separate logger for each connection particularly useful?

Not sure what you mean by that. Each connection uses the same logger. It's the one from the [server](https://github.com/mholt/caddy-l4/blob/c0758c2c40b3daba1a8428da97e5e0db82370b9f/layer4/server.go#L38) that gets passed into `WrapConnection`.  
If you meant what the use case is, I used it mostly for the debug logging in the `Match` function. With this it is possible to see at which matcher layer4 breaks, which was impossible before. My configs often have routes with multiple matcher+handler combos that should match. For example a proxy protocol matcher & handler followed by tls followed by http. So each "layer" gets unwrapped one after another.

**Review by mholt — APPROVED** [review]
Thanks for this. Sorry it took me so long. Let's give it a try and see how it goes! I may ask for your help if anything comes up.

Again, really appreciate it :D

**Comment by ydylla** [comment]
@mholt Thanks for the merge.  
But since you also did merge #78 layer4 will not build right now. Here is the diff to fix it:
```diff
diff --git a/layer4/listener.go b/layer4/listener.go
index 5001d6b..49a6d9f 100644
--- a/layer4/listener.go
+++ b/layer4/listener.go
@@ -119,7 +119,7 @@ func (l *listener) handle(conn net.Conn) {
 	buf.Reset()
 	defer bufPool.Put(buf)
 
-	cx := WrapConnection(conn, buf)
+	cx := WrapConnection(conn, buf, l.logger)
 	cx.Context = context.WithValue(cx.Context, listenerCtxKey, l)
 
 	start := time.Now()
```

**Pull request merged**
